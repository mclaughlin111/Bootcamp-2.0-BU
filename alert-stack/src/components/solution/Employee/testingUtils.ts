import { Employee } from "./Employee";

export const testEmployees: Employee[] = [
  {
    id: 1,
    firstName: "Carol",
    surName: "Pugh",
    gender: "female",
    company: "PYRAMIS",
    email: "carolpugh@pyramis.com",
    phone: "+1 (894) 469-2330",
    address: "396 Gilmore Court, Edenburg, American Samoa, 4850",
    team: "CL",
  },
  {
    id: 2,
    firstName: "Camacho",
    surName: "Abbott",
    gender: "male",
    company: "NAXDIS",
    email: "camachoabbott@naxdis.com",
    phone: "+1 (839) 473-3027",
    address: "541 Bay Parkway, Hampstead, Massachusetts, 7398",
    team: "CD&A",
  },
  {
    id: 3,
    firstName: "Cecilia",
    surName: "Herring",
    gender: "female",
    company: "GEEKETRON",
    email: "ceciliaherring@geeketron.com",
    phone: "+1 (866) 584-2623",
    address: "625 Royce Place, Ticonderoga, Maryland, 3541",
    team: "CM&P",
  },
];
