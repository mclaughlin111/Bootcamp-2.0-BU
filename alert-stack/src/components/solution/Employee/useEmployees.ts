import axios from "axios";
import { useEffect, useState } from "react";
import { Employee } from "./Employee";

export const useEmployees = (searchText: string) => {
  const [isLoading, setIsLoading] = useState(true);
  const [employees, setEmployees] = useState<Employee[]>([]);

  useEffect(() => {
    const getEmployees = async () => {
      setIsLoading(true);
      const { data } = await axios.get<Employee[]>(
        `/employees?search=${searchText}`
      );
      setEmployees(data);
      setIsLoading(false);
    };
    getEmployees();
  }, [searchText]);

  return { isLoading, employees };
};
