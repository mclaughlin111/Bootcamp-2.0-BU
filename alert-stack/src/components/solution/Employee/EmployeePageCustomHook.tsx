import { useState } from "react";
import { EmployeeSearch } from "./EmployeeSearch";
import { EmployeesTable } from "./EmployeesTable";
import { useEmployees } from "./useEmployees";

export const EmployeePageCustomHook = () => {
  const [searchText, setSearchText] = useState("");

  const { employees, isLoading } = useEmployees(searchText);

  return (
    <>
      <EmployeeSearch onSubmitSearch={setSearchText} />
      {isLoading ? (
        <div>Is Loading</div>
      ) : (
        <EmployeesTable employees={employees} />
      )}
    </>
  );
};
