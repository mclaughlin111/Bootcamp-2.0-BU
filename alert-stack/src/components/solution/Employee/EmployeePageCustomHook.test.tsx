import { render, screen } from "@testing-library/react";
import * as hooks from "./useEmployees";
import { EmployeePageCustomHook } from "./EmployeePageCustomHook";
import { testEmployees } from "./testingUtils";
import userEvent from "@testing-library/user-event";

describe("handle useEMployee hook data", () => {
  test("Employee page correctly handles empty array of employees and isLoading=false", async () => {
    jest
      .spyOn(hooks, "useEmployees")
      .mockReturnValue({ isLoading: true, employees: [] });
    const { container } = render(<EmployeePageCustomHook />);

    expect(await screen.findByText("Is Loading")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test("Employee page correctly handles array of employees and isLoading=true", async () => {
    jest
      .spyOn(hooks, "useEmployees")
      .mockReturnValue({ isLoading: false, employees: testEmployees });
    const { container } = render(<EmployeePageCustomHook />);

    expect(await screen.findByRole("table")).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});

test("Employee page allows user to change search string", async () => {
  jest
    .spyOn(hooks, "useEmployees")
    .mockReturnValue({ isLoading: false, employees: testEmployees });
  render(<EmployeePageCustomHook />);

  expect(await screen.findByRole("table")).toBeInTheDocument();

  jest.clearAllMocks();

  const useEmployeeMock = jest
    .spyOn(hooks, "useEmployees")
    .mockReturnValue({ isLoading: true, employees: [] });

  userEvent.type(screen.getByRole("textbox"), "searchString{Enter}");

  expect(await screen.findByText("Is Loading")).toBeInTheDocument();
  expect(useEmployeeMock).toHaveBeenCalledTimes(1);
  expect(useEmployeeMock).toHaveBeenCalledWith("searchString");
});
