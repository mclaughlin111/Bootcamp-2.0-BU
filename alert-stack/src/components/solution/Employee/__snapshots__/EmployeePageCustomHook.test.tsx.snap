// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Employee page correctly handles array of employees and isLoading=true 1`] = `
<div>
  <input
    type="text"
  />
  <button>
    Search
  </button>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-11xur9t-MuiPaper-root-MuiTableContainer-root"
  >
    <table
      aria-label="simple table"
      class="MuiTable-root css-quj9j6-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            First Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Last Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Company
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Email
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Phone
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Address
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Team
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Carol
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Pugh
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            PYRAMIS
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            carolpugh@pyramis.com
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            +1 (894) 469-2330
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            396 Gilmore Court, Edenburg, American Samoa, 4850
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            CL
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Camacho
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Abbott
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            NAXDIS
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            camachoabbott@naxdis.com
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            +1 (839) 473-3027
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            541 Bay Parkway, Hampstead, Massachusetts, 7398
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            CD&A
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Cecilia
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Herring
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            GEEKETRON
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            ceciliaherring@geeketron.com
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            +1 (866) 584-2623
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            625 Royce Place, Ticonderoga, Maryland, 3541
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            CM&P
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Employee page correctly handles empty array of employees and isLoading=false 1`] = `
<div>
  <input
    type="text"
  />
  <button>
    Search
  </button>
  <div>
    Is Loading
  </div>
</div>
`;
